x/exp/batch/batch.go

```diff
@@ -0,0 +1,387 @@
+		return fmt.Errorf("%w: %w", errInvalidPart, err)
+	}
+	if res.Request.Context, err = eval.ValueToRecord(be.env.Context); err != nil {
+		return fmt.Errorf("%w: %w", errInvalidPart, err)
+	}
+	res.Values = be.Values
+	batchCompile(be)
+	res.Decision, res.Diagnostic = isAuthorized(be.evalers, be.env)
+	be.callback(res)
+	return nil
+}
+
+func isAuthorized(ps []*idEvaler, env *eval.Env) (types.Decision, types.Diagnostic) {
+	var diag types.Diagnostic
+	var forbids []types.DiagnosticReason
+	var permits []types.DiagnosticReason
+	// Don't try to short circuit this.
+	// - Even though single forbid means forbid
+	// - All policy should be run to collect errors
+	// - For permit, all permits must be run to collect annotations
+	// - For forbid, forbids must be run to collect annotations
+	for _, po := range ps {
+		result, err := po.Evaler.Eval(env)
+		if err != nil {
+			diag.Errors = append(diag.Errors, types.DiagnosticError{PolicyID: po.PolicyID, Position: types.Position(po.Policy.Position), Message: err.Error()})
+			continue
+		}
+		if !result {
+			continue
+		}
+		if po.Policy.Effect == ast.EffectPermit {
+			permits = append(permits, types.DiagnosticReason{PolicyID: po.PolicyID, Position: types.Position(po.Policy.Position)})
+		} else {
+			forbids = append(forbids, types.DiagnosticReason{PolicyID: po.PolicyID, Position: types.Position(po.Policy.Position)})
+		}
+	}
+	if len(forbids) > 0 {
+		diag.Reasons = forbids
+		return types.Deny, diag
+	}
+	if len(permits) > 0 {
+		diag.Reasons = permits
+		return types.Allow, diag
+	}
+	return types.Deny, diag
+}
+
+// func testPrintPolicy(p *ast.Policy) {
+// 	pp := (*parser.Policy)(p)
+// 	var got bytes.Buffer
+// 	pp.MarshalCedar(&got)
+// 	fmt.Println(got.String())
+// }
+
+func batchCompile(be *batchEvaler) {
+	if be.compiled {
+		return
+	}
+	be.evalers = make([]*idEvaler, len(be.policies))
+	for i, p := range be.policies {
+		be.evalers[i] = &idEvaler{PolicyID: p.PolicyID, Policy: p.Policy, Evaler: eval.Compile(p.Policy)}
+	}
+	be.compiled = true
+}
+
+// cloneSub will return a new value if any of its children have changed
+// and signal the change via the boolean
+func cloneSub(r types.Value, k types.String, v types.Value) (types.Value, bool) {
+	switch t := r.(type) {
+	case types.EntityUID:
+		if key, ok := eval.ToVariable(t); ok && key == k {
+			return v, true
+		}
+	case types.Record:
+		cloned := false
+		for kk, vv := range t {
+			if vv, delta := cloneSub(vv, k, v); delta {
+				if !cloned {
+					t = maps.Clone(t) // intentional shallow clone
+					cloned = true
+				}
+				t[kk] = vv
+			}
+		}
+		return t, cloned
+	case types.Set:
+		cloned := false
+		for kk, vv := range t {
+			if vv, delta := cloneSub(vv, k, v); delta {
+				if !cloned {
+					t = slices.Clone(t) // intentional shallow clone
+					cloned = true
+				}
+				t[kk] = vv
+			}
+		}
+		return t, cloned
+	}
+	return r, false
+}
+
+func findVariables(r types.Value, found []types.String) []types.String {
```

Comment:
Mutating input arguments makes me feel a little icky and I doubt the performance here is super critical. Could you just return the slice of variables that are found?

Also, since it seems like you want this to be more of a set, you could use `map[types.String]struct{}` and `maps.Copy` to merge them.

Reply:
I did a thing, tell me if it looks good enough.